File: 145.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    // helper function
    void post(TreeNode *cur, vector<int>& ret) {
        if (cur == nullptr)
            return;
        post(cur->left, ret);
        post(cur->right, ret);
        ret.push_back(cur->val); 
    }
    
    // Recursive method
    vector<int> postorderTraversalR(TreeNode* root) {
        vector<int> ret;
        post(root, ret);
        return ret;
    }

    // Iterative method
    vector<int> postorderTraversalI(TreeNode* root) {
        vector<int> ret;
        TreeNode *cur = root;
        vector<TreeNode *>st;

        while (!st.empty() || cur != nullptr) {
            while (cur != nullptr) {
                st.push_back(cur);
                st.push_back(cur);
                cur = cur->left;
            }

            cur = st.back();
            st.pop_back();

            if (st.size() > 0 && st.back() == cur)
                cur = cur->right;
            else {
                ret.push_back(cur->val);
                cur = nullptr;
            }
        }

        return ret;
    }
};

File: go/145.go
package main

// Definition for a binary tree node.
// type TreeNode struct {
// 	Val   int
// 	Left  *TreeNode
// 	Right *TreeNode
// }

func postorderTraversal(root *TreeNode) []int {
	var ret = make([]int, 0)
	var st = make([]*TreeNode, 0)
	var cur *TreeNode = root

	for len(st) > 0 || cur != nil {
		for cur != nil {
			st = append(st, cur)
			st = append(st, cur)
			cur = cur.Left
		}

		cur = st[len(st)-1]
		st = st[:len(st)-1]

		if len(st) > 0 && st[len(st)-1] == cur {
			cur = cur.Right
		} else {
			ret = append(ret, cur.Val)
			cur = nil
		}
	}

	return ret
}
